v2/tools/generator/internal/codegen/pipeline/apply_kubernetes_resource_interface.go

```diff
@@ -14,43 +14,36 @@ import (
 	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/interfaces"
 )
 
+const ApplyKubernetesResourceInterfaceStageID = "applyKubernetesResourceInterface"
+
 // ApplyKubernetesResourceInterface ensures that every Resource implements the KubernetesResource interface
 func ApplyKubernetesResourceInterface(idFactory astmodel.IdentifierFactory) Stage {
 	return MakeLegacyStage(
-		"applyKubernetesResourceInterface",
+		ApplyKubernetesResourceInterfaceStageID,
 		"Add the KubernetesResource interface to every resource",
 		func(ctx context.Context, definitions astmodel.TypeDefinitionSet) (astmodel.TypeDefinitionSet, error) {
-			skip := make(map[astmodel.TypeName]struct{})
+			updatedDefs := make(astmodel.TypeDefinitionSet)
 
-			result := make(astmodel.TypeDefinitionSet)
-			for typeName, typeDef := range definitions {
-				if _, ok := skip[typeName]; ok {
-					continue
+			for typeName, typeDef := range astmodel.FindResourceDefinitions(definitions) {
+				resource := typeDef.Type().(*astmodel.ResourceType)
+				newResource, err := interfaces.AddKubernetesResourceInterfaceImpls(typeDef, idFactory, definitions)
+				if err != nil {
+					return nil, errors.Wrapf(err, "couldn't implement Kubernetes resource interface for %q", typeName)
 				}
 
-				if resource, ok := typeDef.Type().(*astmodel.ResourceType); ok {
-					newResource, err := interfaces.AddKubernetesResourceInterfaceImpls(typeName, resource, idFactory, definitions)
-					if err != nil {
-						return nil, errors.Wrapf(err, "couldn't implement Kubernetes resource interface for %q", typeName)
-					}
-
-					// this is really very ugly; a better way?
-					if _, ok := newResource.SpecType().(astmodel.TypeName); !ok {
-						// the resource Spec was replaced with a new definition; update it
-						// by replacing the named definition:
-						specName := resource.SpecType().(astmodel.TypeName)
-						result[specName] = astmodel.MakeTypeDefinition(specName, newResource.SpecType())
-						skip[specName] = struct{}{}
-						newResource = newResource.WithSpec(specName)
-					}
-
-					newDef := typeDef.WithType(newResource)
-					result.Add(newDef)
-				} else {
-					result.Add(typeDef)
+				// this is really very ugly; a better way?
+				if _, ok := newResource.SpecType().(astmodel.TypeName); !ok {
+					// the resource Spec was replaced with a new definition; update it
+					// by replacing the named definition:
+					specName := resource.SpecType().(astmodel.TypeName)
+					updatedDefs.Add(astmodel.MakeTypeDefinition(specName, newResource.SpecType()))
+					newResource = newResource.WithSpec(specName)
 				}
```

Comment:
Do we need something like a `Redefine(TypeName, Type)` method on `TypeDefinitionSet` that overwrites any existing definition by that name with a new one? 

The resulting code would resemble this:

```
				// this is really very ugly; a better way?
				if newType, ok := newResource.SpecType().(astmodel.TypeName); !ok {
					// the resource Spec was replaced with a new definition; update it
					// by replacing the named definition:
					specName := resource.SpecType().(astmodel.TypeName)
					updatedDefs.Redefine(specName, newType)
					newResource = newResource.WithSpec(specName)
				}
```

Reply:
At least in this context though, the `updatedDefs` doesn't have this type defined at all so it's not clear what `Redefine` would do in that case, or how it's different than `Add`?

Reply:
I was thinking that maybe a helper method would reduce the number of steps; if we had overloads, I'd suggest `Add(TypeName, Type)` as an overload for `Add(TypeDefinition)`.